src/core/filesystem.cc

```diff
@@ -1086,7 +1088,8 @@ JoinPath(std::initializer_list<std::string> segments)
       // Check if S3 path (s3://<bucket> or s3://<host>:<port>/<bucket>)
       if (!seg.empty() && !seg.rfind("s3://", 0)) {
         re2::RE2 s3_regex(
```

Comment:
Why do we need all this extra logic for bucket? Segments are anything that is just joined by a '/'.

Reply:
The path in case of minio s3 will contain the s3://host:port/bucket_name/rest_of_the_path
This code extracts the host and port number from the path and returns the true path in the s3 bucket.

Reply:
Why is JoinPath doing all that. It should just be joining segments. If one of the segments is S3:// whatever then why do we need to do anything different. Is there some caller of JoinPath that expects this behavior?

Reply:
Yes, one of the callers of JoinPath expects the s3 path i.e. the Tensorflow savedmodel and graphdef model functions. They expect the s3 path without the `host:port`

Reply:
Then can we have a separate filesystem function (like "Localize") that they can use when necessary to strip that part of the path. And put it in the filesystem objects so that the S3 one can use the common s3_regex. Let's not overload JoinPath to do additional work.

Reply:
Fixed with a similar work around directly in DownloadFileFolder